src/Interpreters/Cache/Metadata.h

```diff
@@ -0,0 +1,181 @@
+    };
+
+    CleanupState getCleanupState(const KeyGuard::Lock &) const { return cleanup_state; }
+
+    void addToCleanupQueue(const FileCacheKey & key, const KeyGuard::Lock &);
+
+    void removeFromCleanupQueue(const FileCacheKey & key, const KeyGuard::Lock &);
+
+private:
+    mutable KeyGuard guard;
+    bool created_base_directory = false;
+    CleanupState cleanup_state = CleanupState::NOT_SUBMITTED;
+    CleanupQueue & cleanup_queue;
+};
+
+using KeyMetadataPtr = std::shared_ptr<KeyMetadata>;
+
+struct CleanupQueue
+{
+    friend struct CacheMetadata;
+public:
+    void add(const FileCacheKey & key);
+    void remove(const FileCacheKey & key);
+    size_t getSize() const;
+
+private:
+    bool tryPop(FileCacheKey & key);
+
+    std::unordered_set<FileCacheKey> keys;
+    mutable std::mutex mutex;
+};
+
+struct CacheMetadata : public std::unordered_map<FileCacheKey, KeyMetadataPtr>, private boost::noncopyable
+{
+public:
+    using Key = FileCacheKey;
+
+    explicit CacheMetadata(const std::string & base_directory_) : base_directory(base_directory_) {}
+
+    String getPathInLocalCache(const Key & key, size_t offset, FileSegmentKind segment_kind) const;
+
+    String getPathInLocalCache(const Key & key) const;
+
+    enum class KeyNotFoundPolicy
+    {
+        THROW,
+        CREATE_EMPTY,
+        RETURN_NULL,
+    };
+    LockedKeyMetadataPtr lockKeyMetadata(const Key & key, KeyNotFoundPolicy key_not_found_policy, bool is_initial_load = false);
+
+    LockedKeyMetadataPtr lockKeyMetadata(const Key & key, KeyMetadataPtr key_metadata, bool return_null_if_in_cleanup_queue = false) const;
+
+    using IterateCacheMetadataFunc = std::function<void(const LockedKeyMetadata &)>;
+    void iterate(IterateCacheMetadataFunc && func);
+
+    void doCleanup();
+
+private:
+    void addToCleanupQueue(const Key & key, const KeyGuard::Lock &);
+    void removeFromCleanupQueue(const Key & key, const KeyGuard::Lock &);
+
+    const std::string base_directory;
+    CacheMetadataGuard guard;
+    CleanupQueue cleanup_queue;
+};
+
+
+/**
+ * `LockedKeyMetadata` is an object which makes sure that as long as it exists the following is true:
+ * 1. the key cannot be removed from cache
+ *    (Why: this LockedKeyMetadata locks key metadata mutex in ctor, unlocks it in dtor, and so
+ *    when key is going to be deleted, key mutex is also locked.
+ *    Why it cannot be the other way round? E.g. that ctor of LockedKeyMetadata locks the key
+ *    right after it was deleted? This case it taken into consideration in createLockedKeyMetadata())
+ * 2. the key cannot be modified, e.g. new offsets cannot be added to key; already existing
+ *    offsets cannot be deleted from the key
+ * And also provides some methods which allow the owner of this LockedKeyMetadata object to do such
+ * modification of the key (adding/deleting offsets) and deleting the key from cache.
+ */
+struct LockedKeyMetadata : private boost::noncopyable
+{
+    LockedKeyMetadata(
```

Comment:
It's confusing to have key and key_metadata as separate parameters. I'd suggest storing key in `KeyMetadata` and use it

Reply:
The comment is still actual

Reply:
I did not say that I fixed it :) if I did, I would have marked it as resolved
I'll ping you when this is ready for the next review